src/session/model/Session.ts

```diff
@@ -0,0 +1,127 @@
+import { Either, Left, Right } from "purify-ts";
+import { AccessTokenMissingError, ExpiresMissingError, SessionExpiredError, SignInInfoMissingError } from "../../error/ErrorFunctions";
+import { Failure } from "../../error/FailureType";
+import { SessionKeys } from "../SessionKeys";
+import { AccessToken } from "./AccessToken";
+import { IMap } from "./ISignInInfo";
+import { Cookie } from "./Cookie";
+import { SessionHandlerConfig } from "../../SessionHandlerConfig";
+
+export class Session {
+
+    public data: IMap<any> = {};
+
+    public constructor(data?: any) {
+
+        if (data) {
+
+            Session.marshall(this, data);
+        }
+
+    }
+
+
+    public unmarshall = () => {
+
+        const obj: any = {};
+        const thisObj: any = this.data;
+
+        const keys = Object.keys(thisObj).sort();
+        for (const i in keys) {
+            if (thisObj.hasOwnProperty(keys[i])) {
+                obj[keys[i]] = thisObj[keys[i]];
+
+            }
+        }
+
+        return obj;
+    };
+
+    public verify = (): Either<Failure, VerifiedSession> => {
+        return VerifiedSession.verifySession(this);
+    };
+
+    public static marshall = (session: Session, data: any): void => {
+        const keys = Object.keys(data).sort();
+
+        for (const i in keys) {
+            if (data.hasOwnProperty(keys[i])) {
+                session.data[keys[i]] = data[keys[i]];
+            }
+        }
+    };
+
+}
+
+export class VerifiedSession extends Session {
+
+    private constructor(session: Session) {
+        super();
+        this.data = session.data;
+    }
+
+    public asCookie = (): Cookie => {
+        return Cookie.sessionCookie(this);
+    };
+
+    public static createNewVerifiedSession(
+        config: SessionHandlerConfig,
+        extraData?: any): VerifiedSession {
+
+        const newCookie: Cookie = Cookie.newCookie(config.cookieSecret);
+
+        const accessToken = AccessToken.createDefaultAccessToken(config.defaultSessionExpiration);
+
+        const signInInfo = {
+            [SessionKeys.AccessToken]: accessToken,
+            [SessionKeys.SignedIn]: 0
+        };
+
+        const sessionData = !extraData ? {} : extraData;
+
+        sessionData[SessionKeys.Id] = newCookie.sessionId;
+        sessionData[SessionKeys.ClientSig] = newCookie.signature;
+        sessionData[SessionKeys.SignInInfo] = signInInfo;
+        sessionData[SessionKeys.Expires] = Date.now() + accessToken.expires_in*1000;
+
+        return new VerifiedSession(new Session(sessionData));
+    }
+
+    public static verifySession(session: Session): Either<Failure, VerifiedSession> {
+
+        const signInInfo = session.data[SessionKeys.SignInInfo];
```

Comment:
Is that even possible to have session without 'sign in' info? Can access token be missing when accounts created session? Can there be session without expires?